File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    
    private void dfs(char[][] grid, int sr, int sc) {
        
        if(sr < 0 || sc < 0 || sr > grid.length -1 || sc > grid[0].length-1 || grid[sr][sc] == '0' || grid[sr][sc] == '2')
            return;
        
        grid[sr][sc] = '2';
        
        dfs(grid, sr + 1, sc);
        dfs(grid, sr, sc + 1);
        dfs(grid, sr - 1, sc);
        dfs(grid, sr, sc - 1);
    }
    
    public int numIslands(char[][] grid) {
        
        int islands = 0;
        for(int i = 0; i < grid.length; i++) {
            
            for(int j = 0; j < grid[0].length; j++) {
                
                if(grid[i][j] == '1') {
                    dfs(grid, i, j);
                    islands++;                    
                }
                
            }
            
        }
        
        return islands;
    }
}

File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    void numIslands(int r, int c, vector<vector<char>> &grid) {
        if(r < 0 || c < 0 || r == grid.size() || c == grid[0].size() || grid[r][c] == '0')
            return;
        
        grid[r][c] = '0';
        
        numIslands(r + 1, c, grid);
        numIslands(r - 1, c, grid);
        numIslands(r, c + 1, grid);
        numIslands(r, c - 1, grid);
    }
    
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        for(int i = 0; i < grid.size(); i++)
            for(int j = 0; j < grid[0].size(); j++)
                if(grid[i][j] == '1') {
                    count++;
                    numIslands(i, j, grid);
                }
        
        return count;
    }
};